### PR TITLE
Add axis source to pointer events

### LINKED ISSUE #2180: Implement axis source pointer events
Some clients misbehave without `wl_pointer.axis_source` events from the mouse wheel. (See https://forum.snapcraft.io/t/the-quest-for-building-a-plain-wayland-electron-kiosk/26739/19)

The information isn't:

1. queried from libinput, nor from X11 or Wayland input platforms;
2. carried by the MirPointerEvents; nor,
3. sent to the client

All of the above should be fixed.

### PR DESCRIPTION
**Pull request by AlanGriffiths**
Fixes #2180

### CODE REVIEW

File: tests/unit-tests/input/evdev/test_libinput_device.cpp [hunk 1]
```diff
@@ -102,6 +94,14 @@ struct MockEventBuilder : mi::EventBuilder
...
                                                                     hscroll, vscroll, relative_x, relative_y);
                                   }));
+
+        ON_CALL(*this, pointer_axis_event(_, _, _, _, _, _, _, _, _, _))
+            .WillByDefault(Invoke([this](MirPointerAxisSource axis_source, std::optional<Timestamp> time, MirPointerAction action, MirPointerButtons buttons, float x,
+                                         float y, float hscroll, float vscroll, float relative_x, float relative_y)
+                                  {
+                                       return builder.pointer_axis_event(axis_source, time, action, buttons, x, y,
+                                                                    hscroll, vscroll, relative_x, relative_y);
+                                  }));
```
Comment (wmww) [inline tests/unit-tests/input/evdev/test_libinput_device.cpp:104]: Some long lines here, this whole thing could use a reflow

File: src/platforms/evdev/libinput_device.cpp [hunk 2]
```diff
@@ -288,8 +287,33 @@ mir::EventUPtr mie::LibInputDevice::convert_axis_event(libinput_event_pointer* p
...
+            relative_x_value, relative_y_value);
+
+    case LIBINPUT_POINTER_AXIS_SOURCE_WHEEL_TILT:
+        return builder->pointer_axis_event(
+            mir_pointer_axis_source_wheel_tilt, time, action, button_state, 0, 0, hscroll_value, vscroll_value,
+            relative_x_value, relative_y_value);
+
+    default:
+        return builder->pointer_event(time, action, button_state, hscroll_value, vscroll_value, relative_x_value,
+                                      relative_y_value);
```
Comment (wmww) [inline src/platforms/evdev/libinput_device.cpp:315]: Lot of code duplication going on here. Can probably stick the axis source in a variable.

File: src/platforms/wayland/input_device.cpp [hunk 3]
```diff
@@ -236,6 +236,31 @@ void miw::PointerInputDevice::pointer_motion(
...
+    std::chrono::nanoseconds event_time,
+    mir::geometry::PointF const& pos,
+    mir::geometry::DisplacementF const& scroll)
+{
+    enqueue([=](EventBuilder* b)
+    {
+        auto const movement = pos - cached_pos;
+        cached_pos = pos;
+        return b->pointer_axis_event(
+            MirPointerAxisSource(pointer_axis_source+1),
```
Comment (wmww) [inline src/platforms/wayland/input_device.cpp:250]: I think this is worth doing a proper switch statement for

File: src/platforms/x11/input/input_device.cpp [hunk 4]
```diff
@@ -228,6 +241,6 @@ void mix::XInputDevice::pointer_motion(
             scroll.dy.as_value(),
             movement.dx.as_value(),
             movement.dy.as_value()
-            )
+           )
```
Comment (wmww) [inline src/platforms/x11/input/input_device.cpp:244]: Whitespace error, perens should probs go on previous line.

File: src/platforms/x11/input/input_device.cpp [hunk 5]
```diff
@@ -201,24 +193,45 @@ void mix::XInputDevice::pointer_release(
...
+            0,
             movement.dx.as_value(),
             movement.dy.as_value()
             )
         );
 
 }
 
-void mix::XInputDevice::pointer_motion(
+void mix::XInputDevice::pointer_motion(EventTime event_time, mir::geometry::PointF pos)
```
Comment (wmww) [inline src/platforms/x11/input/input_device.cpp:205]: It looks to me like code duplication could be removed by this function delegating to `pointer_axis_motion()` with `mir_pointer_axis_source_none` and `0, 0` for scroll.

Perhaps this function could even be removed and callers could just send those arguments to `pointer_axis_motion()`? Don't feel strongly about that either way.

**Review by wmww — CHANGES_REQUESTED** [review]
The changes to the event and overall idea look good.

**Review by RAOF — APPROVED** [review]
All William's requests look sane to me. With them, +1.

File: src/platforms/x11/input/input_device.cpp (hunk 5 above)
Comment (AlanGriffiths) [reply src/platforms/x11/input/input_device.cpp:205]: I don't see how this suggestion improves anything. (At least not without a wider rework of the event builder stack which deserves its own PR!)

File: tests/unit-tests/input/evdev/test_libinput_device.cpp [hunk 6]
```diff
@@ -102,6 +94,18 @@ struct MockEventBuilder : mi::EventBuilder
                                        return builder.pointer_event(time, action, buttons, x, y,
                                                                     hscroll, vscroll, relative_x, relative_y);
                                   }));
+
+        ON_CALL(*this, pointer_axis_event(_, _, _, _, _, _, _, _, _, _))
+            .WillByDefault(
+                Invoke(
```
Comment (Saviq) [inline tests/unit-tests/input/evdev/test_libinput_device.cpp:100]: To reduce clutter, you could drop the `Invoke()` from here. Rest is fine AFAICT.

**Review by Saviq — APPROVED** [review]
bors d+

**Comment by AlanGriffiths** [comment]
I'm landing this, any further issues can go in a follow-up.

bors r=@RAOF,@Saviq

**Pull request merged**